xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiHibernateVersioningStore.java

```diff
@@ -147,11 +151,69 @@ private XWikiDocumentArchive getXWikiDocumentArchive(XWikiDocument doc, Revision
         XWikiContext inputxcontext) throws XWikiException
     {
         XWikiDocumentArchive archiveDoc = doc.getDocumentArchive();
-        // We only retrieve a cached archive if we want a complete one.
-        if (archiveDoc != null && criteria.isAllInclusive()) {
-            return archiveDoc;
+        if (archiveDoc == null) {
+            archiveDoc = getXWikiDocumentArchiveFromDatabase(doc, criteria, inputxcontext);
+        } else if (!criteria.isAllInclusive()) {
+            archiveDoc = filterArchiveFromCriteria(doc, archiveDoc, criteria);
         }
+        return archiveDoc;
+    }
+
+    private XWikiDocumentArchive filterArchiveFromCriteria(XWikiDocument doc, XWikiDocumentArchive archiveDoc,
```

Comment:
To be transparent I mostly trusted the original code that was in https://github.com/xwiki/xwiki-platform/pull/2925/files#diff-bb23b7b80163e5c6cac4ba4592802e30d01ac1d463466454d47e8eaa5bf021d9L2678-L2716 for the logic here, so might be good to review it properly.

Reply:
Since this comment I changed the logic to reuse existing implementation from the interface. Which is backed by some unit tests.